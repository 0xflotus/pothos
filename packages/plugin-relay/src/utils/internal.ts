import { SchemaTypes } from '@pothos/core';
import { decodeGlobalID, encodeGlobalID } from './global-ids';

export function internalEncodeGlobalID<Types extends SchemaTypes>(
  builder: PothosSchemaTypes.SchemaBuilder<Types>,
  typename: string,
  id: bigint | number | string,
) {
  if (builder.options.relayOptions.encodeGlobalID) {
    return builder.options.relayOptions.encodeGlobalID(typename, id);
  }

  return encodeGlobalID(typename, id);
}

export function internalDecodeGlobalID<Types extends SchemaTypes>(
  builder: PothosSchemaTypes.SchemaBuilder<Types>,
  globalID: string,
) {
  if (builder.options.relayOptions.decodeGlobalID) {
    return builder.options.relayOptions.decodeGlobalID(globalID);
  }

  return decodeGlobalID(globalID);
}
