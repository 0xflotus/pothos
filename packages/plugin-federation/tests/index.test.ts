// import { gql } from 'apollo-server';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
// import { User } from './examples/simple/data';
import exampleSchema from './examples/simple/schema';

describe('federation example schema', () => {
  it('generates expected schema', () => {
    expect(printSchemaWithDirectives(exampleSchema)).toMatchSnapshot();
  });

  // describe('queries', () => {
  //   it('query some stuff', async () => {
  //     const query = gql`
  //       query {
  //         user(id: 1) {
  //           id
  //           firstName
  //           lastName
  //           contactInfo {
  //             email
  //             phoneNumber
  //           }
  //         }
  //       }
  //     `;

  //     const result = await execute({
  //       schema: exampleSchema,
  //       document: query,
  //       contextValue: {
  //         User,
  //       },
  //     });

  //     expect(result).toMatchSnapshot();
  //   });
  // });
});
