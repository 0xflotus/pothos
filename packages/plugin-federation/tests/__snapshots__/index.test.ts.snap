// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`federation accounts schema generates expected schema 1`] = `
"type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  me: User
}

type User @key(fields: \\"id\\") {
  id: ID
  name: String
  username: String
}
"
`;

exports[`federation accounts schema generates expected schema 2`] = `
"directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @key(fields: String!, resolvable: Boolean = true) on INTERFACE | OBJECT

directive @link(import: [link__Import], url: String!) on SCHEMA

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  me: User
}

type User {
  id: ID
  name: String
  username: String
}

scalar _Any

union _Entity = User

type _Service {
  \\"\\"\\"
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  \\"\\"\\"
  sdl: String
}

scalar link__Import"
`;

exports[`federation inventory schema generates expected schema 1`] = `
"type Product @key(fields: \\"upc\\") @extends {
  inStock: Boolean
  price: Float @external
  shippingEstimate: Float @requires(fields: \\"price weight\\")
  upc: String
  weight: Float @external
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}
"
`;

exports[`federation inventory schema generates expected schema 2`] = `
"directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @key(fields: String!, resolvable: Boolean = true) on INTERFACE | OBJECT

directive @link(import: [link__Import], url: String!) on SCHEMA

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

type Product {
  inStock: Boolean
  price: Float
  shippingEstimate: Float
  upc: String
  weight: Float
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

scalar _Any

union _Entity = Product

type _Service {
  \\"\\"\\"
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  \\"\\"\\"
  sdl: String
}

scalar link__Import"
`;

exports[`federation products schema generates expected schema 1`] = `
"type Product @key(fields: \\"upc\\") {
  name: String
  price: Float
  upc: String
  weight: Float
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  topProducts(first: Int = 5): [Product!]
}
"
`;

exports[`federation products schema generates expected schema 2`] = `
"directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @key(fields: String!, resolvable: Boolean = true) on INTERFACE | OBJECT

directive @link(import: [link__Import], url: String!) on SCHEMA

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

type Product {
  name: String
  price: Float
  upc: String
  weight: Float
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
  topProducts(first: Int = 5): [Product!]
}

scalar _Any

union _Entity = Product

type _Service {
  \\"\\"\\"
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  \\"\\"\\"
  sdl: String
}

scalar link__Import"
`;

exports[`federation reviews schema generates expected schema 1`] = `
"type Product @key(fields: \\"upc\\") @extends {
  reviews: [Review!]
  upc: String
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

type Review @key(fields: \\"id\\") {
  author: User @provides(fields: \\"username\\")
  body: String
  id: ID
  product: Product
}

type User @key(fields: \\"id\\") @extends {
  id: ID
  reviews: [Review!]
  username: String @external
}
"
`;

exports[`federation reviews schema generates expected schema 2`] = `
"directive @extends on INTERFACE | OBJECT

directive @external on FIELD_DEFINITION | OBJECT

directive @key(fields: String!, resolvable: Boolean = true) on INTERFACE | OBJECT

directive @link(import: [link__Import], url: String!) on SCHEMA

directive @provides(fields: String!) on FIELD_DEFINITION

directive @requires(fields: String!) on FIELD_DEFINITION

directive @shareable on FIELD_DEFINITION | OBJECT

type Product {
  reviews: [Review!]
  upc: String
}

type Query {
  _entities(representations: [_Any!]!): [_Entity]!
  _service: _Service!
}

type Review {
  author: User
  body: String
  id: ID
  product: Product
}

type User {
  id: ID
  reviews: [Review!]
  username: String
}

scalar _Any

union _Entity = Product | Review | User

type _Service {
  \\"\\"\\"
  The sdl representing the federated service capabilities. Includes federation directives, removes federation types, and includes rest of full schema after schema directives have been applied
  \\"\\"\\"
  sdl: String
}

scalar link__Import"
`;
